### ISSUE TITLE
Use Video Toolbox to decode VPx and AV1 videos on macOS 11+

### ISSUE THREAD

**Issue by arafey** [iina/iina#3241]
**What you want IINA to do:**
Use Video Toolbox's native VPx/AV1 decoding to take advantage of hardware decoding on compatible macOS devices (more recent Intel devices for VP8/9 and only Apple Silicon devices for AV1).

**What IINA does currently:**
No Video Toolbox decoding of VPx/AV1 playback, always uses libmpv fallback (as of IINA 1.1.2 Build 128)

![vp9_iina_inspector](https://user-images.githubusercontent.com/7505951/101436478-3cdcdc80-38dc-11eb-8f69-ddbf7c73ac63.png)

**Why you think this should be added:**
IINA already supports Video Toolbox and hardware decoding. Hardware decoding of these codecs is particularly useful because they're widely used for UHD/HDR/HFR videos (e.g. almost all high quality YouTube videos) and their decoders are very complex. Playback of a high resolution, high framerate, high dynamic range video using a complex decoder = high CPU utilisation (high power consumption, hot temps, loud fans, bad performance, etc). As codecs become more complex, the need for hardware decoding becomes more vital for proper playback without excessive power consumption on battery-powered devices like laptops.

I ran a brief test comparing VP9 software and hardware decoding to illustrate the significant difference. The results with AV1 are likely to be more extreme as it's even more complex.

**Examples of other projects that have something similar:**
- Google Chrome v86 (see test results below)
- Safari
- QuickTime Player
- possibly others

**Test Results**:

_Intel Power Gadget: CPU Package, CPU Core, GPU, and DRAM Power Consumption and Utilisation (First Run)_
Idle | Chrome (HW) | Chrome (SW) | IINA (SW)
:--:|:--:|:--:|:--:
![idle_power](https://user-images.githubusercontent.com/7505951/101436335-f8514100-38db-11eb-8689-712b8cd04d6f.png) | ![vp9_hwdec_power](https://user-images.githubusercontent.com/7505951/101436349-0010e580-38dc-11eb-821d-968ce07529ea.png) | ![vp9_swdec_power](https://user-images.githubusercontent.com/7505951/101437078-6d714600-38dd-11eb-9319-da41b8bcde70.png) | ![vp9_iina_power](https://user-images.githubusercontent.com/7505951/101436398-1323b580-38dc-11eb-9fbe-abfb1bbd69c0.png)

_Activity Monitor: CPU Core Utilisation (Second Run)_
Idle | Chrome (HW) | Chrome (SW) | IINA (SW)
:--:|:--:|:--:|:--:
![idle_cpu](https://user-images.githubusercontent.com/7505951/101438454-67c92f80-38e0-11eb-87ce-1fc8f6fdc3e5.png) | ![vp9_hwdec_cpu](https://user-images.githubusercontent.com/7505951/101438456-67c92f80-38e0-11eb-9402-0799b31be4e9.png) | ![vp9_swdec_cpu](https://user-images.githubusercontent.com/7505951/101438459-6861c600-38e0-11eb-8b30-5e9700521cad.png) | ![vp9_iina_cpu](https://user-images.githubusercontent.com/7505951/101438458-6861c600-38e0-11eb-8e2a-5199cba2b63f.png)

_Activity Monitor: Integrated and Discrete GPU Utilisation (Second Run)_
Idle | Chrome (HW) | Chrome (SW) | IINA (SW)
:--:|:--:|:--:|:--:
![idle_gpu](https://user-images.githubusercontent.com/7505951/101438546-a65eea00-38e0-11eb-8354-80816ae3aeda.png) | ![vp9_hwdec_gpu](https://user-images.githubusercontent.com/7505951/101438547-a65eea00-38e0-11eb-868e-eab1759a03f4.png) | ![vp9_swdec_gpu](https://user-images.githubusercontent.com/7505951/101438549-a6f78080-38e0-11eb-998c-519f60e26b19.png) | ![vp9_iina_gpu](https://user-images.githubusercontent.com/7505951/101438548-a6f78080-38e0-11eb-9d2e-ae3cbf4cec39.png)

**System and Test Information:**

| Model | [MacBook Pro (16-inch, 2019)](https://support.apple.com/kb/SP809) |
| :--- | :---: |
| CPU | Intel Core i9-9880H |
| Integrated GPU | Intel UHD Graphics 630 |
| Discrete GPU | AMD Radeon Pro 5500M 8GB GDDR6 |
| Memory | 32GB DDR4-2666 (16GB×2) |

Chrome requires the videotoolbox-vp9-decoding flag to be enabled.

![vt-vp9dec-enabled](https://user-images.githubusercontent.com/7505951/101436459-30f11a80-38dc-11eb-883d-e879509f4401.png)

I don't have access to an Apple Silicon macOS device, so I could not test AV1 HW support. The idle test was run with only Intel Power Gadget open during the first run and only Activity Monitor during the second run. The decoding tests were run while playing [this video](https://youtu.be/mkggXE5e2yk), which was first downloaded locally as a 4K60 SDR VP9 + Opus file using [youtube-dl](https://github.com/ytdl-org/youtube-dl):
```
youtube-dl -f 315+251 "https://youtu.be/mkggXE5e2yk"
```
This issue is a duplicate of #1179 but that issue was opened and closed prior to macOS 11 and the addition of Video Toolbox VPx/AV1 support.

Please let me know if there's any other info or if there are any questions.

**Comment by anohren** [comment]
That's a nice looking report.

Sorry in advance for not having anything better to add than: interesting to know those machines have hardware AV1 support

**Comment by arafey** [comment]
> Sorry in advance for not having anything better to add than: interesting to know those machines have hardware AV1 support

@anohren No need to be sorry. I figured the dev(s) already know most of what I said but wanted to be more informative for other users that ended up reading it.

**Comment by galad87** [comment]
VideToolbox does not expose a AV1 decoder yet. So even if M1 SoC could include one, it can't be accessed yet.

**Comment by arafey** [comment]
> VideToolbox does not expose a AV1 decoder yet. So even if M1 SoC could include one, it can't be accessed yet.

That's a shame. I assumed it did because of [this](https://www.cpu-monkey.com/en/igpu-apple_m1_8_core-224). Where can you see which codecs are supported by Video Toolbox?

**Comment by junzhli** [comment]
Hi! Have the same question: is it possible to add support for VP9 decoding?

**Comment by wjohhan** [comment]
Movist support VP9 Hardware decoding, IDK how did he do it.

**Comment by samueldemoura** [comment]
Seems like VideoToolbox is already exposing VP9 and AV1 decoders (at least as per [Apple's documentation](https://developer.apple.com/documentation/coremedia/1564239-video_codec_constants)), but out of those, FFmpeg [only supports VP9 on master right now](https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/videotoolbox.c#L907). Doesn't look like VP8 is supported at all.

Can anyone more familiar with the codebase chime in to let us know how big of an effort it would be to get hardware-decoded VP9 in IINA, and if they need any help with this?

**Comment by vade** [comment]
It seems like you can

```
        VTRegisterSupplementalVideoDecoderIfAvailable(kCMVideoCodecType_AV1)
        VTRegisterSupplementalVideoDecoderIfAvailable(kCMVideoCodecType_VP9)

```

And it should just work, assuming that there is AV1 or VP9 sample buffers that are exposed in a way that can be fetched / demuxed easily. Its not usual to see those in MP4 or MOV containers which is typically how AVFoundation or Video Toolbox would get those frames, iirc? 


And yea, this is a great bug report! Well written!

**Comment by low-batt** [comment]
The next release of IINA will include upgrading [mpv](https://mpv.io/) to 0.35.1 and [FFmpeg](https://ffmpeg.org/) to 6.0. I have confirmed these upgrades provide support for hardware decoding of videos encoded with VP9 and Apple ProRes codecs.

On the question of hardware decoding for videos encoded using AV1…

The layering is:
- [IINA](https://iina.io/)
- [mpv](https://mpv.io/) 
- [FFmpeg](https://ffmpeg.org/) 
- [Video Toolbox](https://developer.apple.com/documentation/videotoolbox)
- Hardware (media engine)

With the last 3 being the critical parts.

I'm having a hard time finding reliable information about what exactly Apple supports. So I coded a quick test that calls this method:
```swift
func printHardwareDecodeStatus(_ codeName: String, _ codecType: CMVideoCodecType) {
  VTRegisterSupplementalVideoDecoderIfAvailable(codecType)
  print("\(codeName) \(VTIsHardwareDecodeSupported(codecType))")
}
```

On all of the [Video Codec Constants](https://developer.apple.com/documentation/coremedia/1564239-video_codec_constants). Full output below, but what we care about is:

```text
kCMVideoCodecType_VP9 true
kCMVideoCodecType_AV1 false
```

I ran this test on my MacBookPro18,2 with the M1 Max chip under macOS Ventura 13.4.

Thus, when it comes to hardware decoding for AV1 we are currently waiting on Apple to implement it, assuming the needed hardware capabilities are available in the current media engine.

Note that without the call to `VTRegisterSupplementalVideoDecoderIfAvailable`, Video Toolbox will report VP9 hardware decoding is not supported.

<details>
<summary><b>Test run:</b></summary>

```text
low-batt@gag Debug$ ./Decoders
kCMVideoCodecType_422YpCbCr8 false
kCMVideoCodecType_Animation false
kCMVideoCodecType_Cinepak false
kCMVideoCodecType_JPEG true
kCMVideoCodecType_JPEG_OpenDML true
kCMVideoCodecType_SorensonVideo false
kCMVideoCodecType_SorensonVideo3 false
kCMVideoCodecType_H263 false
kCMVideoCodecType_H264 true
kCMVideoCodecType_HEVC true
kCMVideoCodecType_HEVCWithAlpha true
kCMVideoCodecType_DolbyVisionHEVC true
kCMVideoCodecType_MPEG4Video false
kCMVideoCodecType_MPEG2Video false
kCMVideoCodecType_MPEG1Video false
kCMVideoCodecType_VP9 true
kCMVideoCodecType_DVCNTSC false
kCMVideoCodecType_DVCPAL false
kCMVideoCodecType_DVCProPAL false
kCMVideoCodecType_DVCPro50NTSC false
kCMVideoCodecType_DVCPro50PAL false
kCMVideoCodecType_DVCPROHD720p60 false
kCMVideoCodecType_DVCPROHD720p50 false
kCMVideoCodecType_DVCPROHD1080i60 false
kCMVideoCodecType_DVCPROHD1080i50 false
kCMVideoCodecType_DVCPROHD1080p30 false
kCMVideoCodecType_DVCPROHD1080p25 false
kCMVideoCodecType_AppleProRes4444XQ true
kCMVideoCodecType_AppleProRes4444 true
kCMVideoCodecType_AppleProRes422HQ true
kCMVideoCodecType_AppleProRes422 true
kCMVideoCodecType_AppleProRes422LT true
kCMVideoCodecType_AppleProRes422Proxy true
kCMVideoCodecType_AppleProResRAW true
kCMVideoCodecType_AppleProResRAWHQ true
kCMVideoCodecType_DisparityHEVC true
kCMVideoCodecType_DepthHEVC true
kCMVideoCodecType_AV1 false
low-batt@gag Debug$ 
```
</details>

**Comment by jadenquinn** [comment]
> **What you want IINA to do:** Use Video Toolbox's native VPx/AV1 decoding to take advantage of hardware decoding on compatible macOS devices (more recent Intel devices for VP8/9 and only Apple Silicon devices for AV1).
> 
> **What IINA does currently:** No Video Toolbox decoding of VPx/AV1 playback, always uses libmpv fallback (as of IINA 1.1.2 Build 128)
> 
> ![vp9_iina_inspector](https://user-images.githubusercontent.com/7505951/101436478-3cdcdc80-38dc-11eb-8f69-ddbf7c73ac63.png)
> 
> **Why you think this should be added:** IINA already supports Video Toolbox and hardware decoding. Hardware decoding of these codecs is particularly useful because they're widely used for UHD/HDR/HFR videos (e.g. almost all high quality YouTube videos) and their decoders are very complex. Playback of a high resolution, high framerate, high dynamic range video using a complex decoder = high CPU utilisation (high power consumption, hot temps, loud fans, bad performance, etc). As codecs become more complex, the need for hardware decoding becomes more vital for proper playback without excessive power consumption on battery-powered devices like laptops.
> 
> I ran a brief test comparing VP9 software and hardware decoding to illustrate the significant difference. The results with AV1 are likely to be more extreme as it's even more complex.
> 
> **Examples of other projects that have something similar:**
> 
> * Google Chrome v86 (see test results below)
> * Safari
> * QuickTime Player
> * possibly others
> 
> **Test Results**:
> 
> _Intel Power Gadget: CPU Package, CPU Core, GPU, and DRAM Power Consumption and Utilisation (First Run)_
> 
> Idle	Chrome (HW)	Chrome (SW)	IINA (SW)
> ![idle_power](https://user-images.githubusercontent.com/7505951/101436335-f8514100-38db-11eb-8689-712b8cd04d6f.png)	![vp9_hwdec_power](https://user-images.githubusercontent.com/7505951/101436349-0010e580-38dc-11eb-821d-968ce07529ea.png)	![vp9_swdec_power](https://user-images.githubusercontent.com/7505951/101437078-6d714600-38dd-11eb-9319-da41b8bcde70.png)	![vp9_iina_power](https://user-images.githubusercontent.com/7505951/101436398-1323b580-38dc-11eb-9fbe-abfb1bbd69c0.png)
> ###### HW decoding using the integrated GPU uses less power than SW decoding using the CPU (above).
> _Activity Monitor: CPU Core Utilisation (Second Run)_
> 
> Idle	Chrome (HW)	Chrome (SW)	IINA (SW)
> ![idle_cpu](https://user-images.githubusercontent.com/7505951/101438454-67c92f80-38e0-11eb-87ce-1fc8f6fdc3e5.png)	![vp9_hwdec_cpu](https://user-images.githubusercontent.com/7505951/101438456-67c92f80-38e0-11eb-9402-0799b31be4e9.png)	![vp9_swdec_cpu](https://user-images.githubusercontent.com/7505951/101438459-6861c600-38e0-11eb-8b30-5e9700521cad.png)	![vp9_iina_cpu](https://user-images.githubusercontent.com/7505951/101438458-6861c600-38e0-11eb-8e2a-5199cba2b63f.png)
> ###### The drop in CPU utilisation with HW decoding (above) results in only a correspondingly small increase in integrated GPU utilisation (below).
> _Activity Monitor: Integrated and Discrete GPU Utilisation (Second Run)_
> 
> Idle	Chrome (HW)	Chrome (SW)	IINA (SW)
> ![idle_gpu](https://user-images.githubusercontent.com/7505951/101438546-a65eea00-38e0-11eb-8354-80816ae3aeda.png)	![vp9_hwdec_gpu](https://user-images.githubusercontent.com/7505951/101438547-a65eea00-38e0-11eb-868e-eab1759a03f4.png)	![vp9_swdec_gpu](https://user-images.githubusercontent.com/7505951/101438549-a6f78080-38e0-11eb-998c-519f60e26b19.png)	![vp9_iina_gpu](https://user-images.githubusercontent.com/7505951/101438548-a6f78080-38e0-11eb-9d2e-ae3cbf4cec39.png)
> **System and Test Information:**
> 
> Model	[MacBook Pro (16-inch, 2019)](https://support.apple.com/kb/SP809)
> CPU	Intel Core i9-9880H
> Integrated GPU	Intel UHD Graphics 630
> Discrete GPU	AMD Radeon Pro 5500M 8GB GDDR6
> Memory	32GB DDR4-2666 (16GB×2)
> Chrome requires the videotoolbox-vp9-decoding flag to be enabled.
> 
> ![vt-vp9dec-enabled](https://user-images.githubusercontent.com/7505951/101436459-30f11a80-38dc-11eb-883d-e879509f4401.png)
> 
> I don't have access to an Apple Silicon macOS device, so I could not test AV1 HW support. The idle test was run with only Intel Power Gadget open during the first run and only Activity Monitor during the second run. The decoding tests were run while playing [this video](https://youtu.be/mkggXE5e2yk), which was first downloaded locally as a 4K60 SDR VP9 + Opus file using [youtube-dl](https://github.com/ytdl-org/youtube-dl):
> 
> ```
> youtube-dl -f 315+251 "https://youtu.be/mkggXE5e2yk"
> ```
> 
> This issue is a duplicate of #1179 but that issue was opened and closed prior to macOS 11 and the addition of Video Toolbox VPx/AV1 support.
> 
> Please let me know if there's any other info or if there are any questions.

I have pretty much the exact same computer on my lap here (except I have a 5600M instead of the 5500M), and Google Chrome doesn't seem to do hardware decoding with VT at all, it just uses 45-70% of my CPU at all times, which is just horrible, since Safari (and a browser in beta that's basically a better Safari, Orion) use the Apple WebKit and VideoToolbox to decode YouTube videos (and maybe other shizz too), and I'm on MacOS 12. I don't know when this changed, but as I see it here, the VT DOES support VPx hardware decoding, and using WebKit does seem to make whatever is using it decode using VT, which is definietly much better than the seemingly software decoding that Chrome and Firefox do, even the Chrome YouTube app does the same thing, if not worse, since it has to open BOTH Chrome and the YouTube app at the same time, and you cannot close Chrome or it'll also close the YouTube app.. which is really dumb.

**Comment by galad87** [comment]
Chrome probably uses AV1 on YouTube, there should be an option somewhere to disable AV1 decoding, or an extension to disable it on YouTube.
Anyway, I guess that's off topic.

**Comment by oscarbg** [comment]
Update on this now that m3 is out with av1 decode? Included in ffmpeg 7?

**Comment by low-batt** [comment]
The next step is for FFmpeg to address ticket [#10642](https://trac.ffmpeg.org/ticket/10642), "AV1 hardware decoding for Apple M3".

**Comment by ePirat** [comment]
Supported upstream in FFmpeg since https://github.com/FFmpeg/FFmpeg/commit/f9c5c5358cfef3847674c6f3b3ded9611ebc5647 (Not in any release yet)

**Comment by low-batt** [comment]
THANKS for letting us know. I had not noticed FFmpeg ticket [#10642](https://trac.ffmpeg.org/ticket/10642) had closed. We are close to the next release of IINA, which upgrades FFmpeg to 7.0.1. Hopefully this enhancement will have been released by the next time IINA upgrades dependencies.

Support for VP9 hardware acceleration is still disabled in IINA when running on Intel Macs. Your update prompted me to test the status of that problem and I just confirmed the problem is still reproducible under macOS Sequoia. This is FFmpeg ticket [#9599](https://trac.ffmpeg.org/ticket/9599). I guess the patch I posted in [this comment](https://trac.ffmpeg.org/ticket/9599#comment:23) is not the proper way to fix that problem. Hoping to eventually get a fix so that VP9 hardware acceleration can be used on Intel Macs.

**Comment by ePirat** [comment]
I looked into the VP9 issue myself too, the attached patch of yours isn't that suitable for upstream I think. (We should either move probably all VT decoders to proper decoders instead of hwdecs or none, and also it breaks API)

**Comment by low-batt** [comment]
I thought that might be the case. I don't know the FFmpeg code base well enough to propose a proper fix.

I will keep in mind that upgrading FFmpeg will get us AV1 hw support. That is an important enhancement.

**Comment by uiryuu** [comment]
Fixed in https://github.com/iina/iina/pull/6261
